Storage/src/Connection/Rest.php

```diff
@@ -305,8 +307,10 @@ private function resolveUploadOptions(array $args)
         }
 
         // @todo add support for rolling hash
-        if ($args['validate'] && !isset($args['metadata']['md5Hash'])) {
+        if (($args['validate'] === true || $args['validate'] === 'md5') && !isset($args['metadata']['md5Hash'])) {
             $args['metadata']['md5Hash'] = base64_encode(Psr7\hash($args['data'], 'md5', true));
+        } elseif ($args['validate'] === 'crc32' && !isset($args['metadata']['crc32c'])) {
+            $args['metadata']['crc32c'] = $this->crcFromStream($args['data']);
```

Comment:
IIUC, this only works if `$args['data']` follows a File interface?

Reply:
@frankyn crc32 has serious performance drawbacks in situations where the crc32c extension is not installed and where the php installation does not support it out of the box. In other words, when we have to fall back to the pure-PHP implementation, we see unacceptable performance losses:

```
Google\CRC32\PHP	        256	24267       1.24 MB/s
Google\CRC32\PHPSlicedBy4	256	24707       1.26 MB/s
Google\CRC32\Builtin	    256	1582535     81.03 MB/s
Google\CRC32\Google	        256	1812113     92.78 MB/s
```

This corresponds to roughly 87 seconds for a 130mb file in our testing.

@dwsupplee and I were talking about being a bit smarter when choosing a validation algorithm. In situations where the user does not explicitly choose one or the other, I think it should be possible to determine the best option. In PHP 7.4 and up, or when the crc32c extension is installed, we can default to crc32 without losing performance. In other cases, we can default to md5.

WDYT?